apps/shell/tash_command.c

```diff
@@ -233,103 +233,233 @@ static int tash_history(int argc, char **args)
+	#define SAVE_USER_TEMPCMD \
+		do { \
+			cmd_line[(*cmd_char_ptr) + 1] = ASCII_NUL; \
+			while (*cmd_char_ptr >= 0) { \
+				cmd_line[*cmd_char_ptr] = cmd[*cmd_char_ptr]; \
+				(*cmd_char_ptr)--; \
+			} \
+		} while (0)
+
+	if (IS_HIST_NOCMD) {
+		/* There is no command executed, let's ignore. */
+
 		return false;
-	} else {
-		/* Save current command. */
-		while (idx < TASH_LINEBUFLEN && cmd[idx] != 0) {
-			cmd_store[cmd_pos][idx] = cmd[idx];
-			idx++;
-		}
-		cmd_store[cmd_pos][idx] = 0;
 	}
 
-	if (status == ASCII_A) { // up
+	if (UP_KEY_PRESSED) {
+		/* UP key Pressed */
+
+		if (IS_HIST_AT_TOP) {
+			/* Already reached the top of history list (no more command at up direction), let's ignore. */
+
+			return false;
+		} else if (IS_HIST_AT_BOT && !HAS_USER_INPUT) {
+			/* First UP key, Save current user input command in temporary buffer */
+
+			SAVE_USER_TEMPCMD;
+		}
+		/* Get previous command index */
+
 		CMD_INDEX_DOWN(cmd_pos);
-		*pos = 0;
+		/* Copy it into cmd buffer */
 
-		while (*pos < TASH_LINEBUFLEN && cmd_store[cmd_pos][*pos] != 0) {
-			cmd[*pos] = cmd_store[cmd_pos][*pos];
-			(*pos)++;
+		GET_HIST_CMD;
+
+	} else if (DOWN_KEY_PRESSED) {
+		/* DOWN key Pressed */
+
+		if (IS_HIST_AT_BOT) {
+			/* Already reached the bottom of history list (no more command at down direction), let's ignore. */
+
+			return false;
 		}
-	} else if (status == ASCII_B) { // down
+
+		/* Get next command index */
+
 		CMD_INDEX_UP(cmd_pos);
-		*pos = 0;
-		if (cmd_pos == cmd_tail) {
-			while (*pos < TASH_LINEBUFLEN && cmd_line[*pos] != 0) {
-				cmd[*pos] = cmd_line[*pos];
-				(*pos)++;
+		if (IS_HIST_AT_BOT) {
+			if (HAS_USER_INPUT) {
+				/* There is the user input command in temporary buffer, take it. */
+
+				GET_USER_TEMPCMD;
+			} else {
+				/* No command in temporary buffer, let's give empty */
+
+				cmd[0] = ASCII_NUL;
+				*cmd_char_ptr = 0;
 			}
 		} else {
-			while (*pos < TASH_LINEBUFLEN && cmd_store[cmd_pos][*pos] != 0) {
-				cmd[*pos] = cmd_store[cmd_pos][*pos];
-				(*pos)++;
-			}
+			/* Copy stored command into cmd buffer. */
+
+			GET_HIST_CMD;
 		}
+
+	} else {
+		shdbg("Not supported\n");
+		return false;
 	}
+
 	return true;
 }
+
+/**
+ * @brief     treat exclamation command
+ * @details   Find the exclamation command and replace it to real command into buffer
+ * @param[in] buff The pointer of user input string
+ * @return    OK on success, ERROR on failure
+ */
+
+int check_exclam_cmd(char *buff)
+{
+	int hist_idx;
+	int histcmd_size;
+	char *histcmd_ptr;
+	char *exclam_ptr;
+	int ret = OK;
+
+	/* Find the '!' in the input character */
+
+	while ((exclam_ptr = strchr(buff, ASCII_EXCLAM)) != NULL) {
+		/* Get the command from history according to the given index */
+
+		hist_idx = strtol(exclam_ptr + 1, &buff, 10);
+		histcmd_ptr = tash_get_cmd_from_history(hist_idx, NULL);
+		if (!histcmd_ptr) {
+			/* No command, Let's finish */
+
+			ret = ERROR;
```

Comment:
If histcmd_ptr was valid in previous loop, it still return ERROR. is it acceptable?
If not initialize ret as ERROR, then set as OK in else part will be better, If I am right.

Reply:
It should return ERROR when it meets a failure regardless of previous loop result.
You can find the same result on Linux.
When history has upto 100 index, let's type `!99;!101`. Then it does not execute `!99` even it is valid.

Reply:
okie understood.